downstairs/src/extent_inner_raw_v2.rs

```diff
@@ -0,0 +1,1058 @@
+            )
+            .map_err(|e| {
+                CrucibleError::IoError(format!(
+                    "extent {}: write failed: {e}",
+                    self.extent_number
+                ))
+            });
+            match r {
+                Err(e) => err = Some(e),
+                Ok(num_bytes) => total_bytes += num_bytes,
+            };
+        }
+
+        let r = match err {
+            Some(e) => Err(e),
+            None if total_bytes != expected_bytes => {
+                Err(CrucibleError::IoError(format!(
+                    "extent {}: incomplete write \
+                    (expected {expected_bytes}, got {total_bytes})",
+                    self.extent_number
+                )))
+            }
+            None => Ok(()),
+        };
+
+        if r.is_err() {
+            for i in 0..n_blocks {
+                let block = BlockOffset(write.offset.0 + i as u64);
+
+                // Try to recompute the context slot from the file.  If this
+                // fails, then we _really_ can't recover, so bail out
+                // unceremoniously.
+                self.recompute_block_written_from_file(block).unwrap();
+            }
+        } else {
+            // Now that writes have gone through, mark as written
+            self.block_written[(write.offset.0) as usize..][..n_blocks]
+                .fill(true);
+        }
+        cdt::extent__write__file__done!(|| {
+            (job_id.0, self.extent_number.0, n_blocks as u64)
+        });
+
+        r
+    }
+
+    fn read(
+        &mut self,
+        job_id: JobId,
+        req: ExtentReadRequest,
+        iov_max: usize, // unused by raw backend
+    ) -> Result<ExtentReadResponse, CrucibleError> {
+        let mut buf = req.data;
+        let block_size = self.extent_size.block_size_in_bytes() as usize;
+        let start_block = req.offset;
+        let num_blocks = buf.capacity() / block_size;
+        check_input(self.extent_size, start_block, buf.capacity())?;
+
+        let start_pos = self.layout.block_pos(start_block);
+        let mut buf_ptr =
+            buf.spare_capacity_mut().as_mut_ptr() as *mut libc::c_void;
+
+        let mut ctxs =
+            vec![[0u8; BLOCK_CONTEXT_SIZE_BYTES as usize]; num_blocks];
+        let mut padding = vec![0u8; self.layout.padding_size() as usize];
+        let mut iovecs = Vec::with_capacity(num_blocks * 3);
+
+        let mut block = start_block;
+        let mut padding_count = 0;
+        for ctx in &mut ctxs {
+            iovecs.push(libc::iovec {
+                iov_base: buf_ptr,
+                iov_len: block_size,
+            });
+            iovecs.push(libc::iovec {
+                iov_base: ctx as *mut _ as *mut _,
+                iov_len: BLOCK_CONTEXT_SIZE_BYTES as usize,
+            });
+            if self.layout.has_padding_after(block) {
+                iovecs.push(libc::iovec {
+                    iov_base: padding.as_mut_ptr() as *mut _,
```

Comment:
What is preadv's behaviour when it's passed multiple iovecs with the same address?

Reply:
I believe they're processed in array order, so the last one wins.  It doesn't really matter here, because we're not using the padding data for anything.